### PR TITLE
Kubelet: surface the container creation/start failure reason

### PR DESCRIPTION
**Pull request by yujuhong**
Container creation/start failure cannot be reproduced by inspecting the
containers. This change caches such errors so that kubelet can retrieve it
later.

This change also extends FakeDockerClient to support setting error response
for a specific function.

This addresses #6321

### CODE REVIEW

**Comment by yujuhong** [comment]
Caveat: If the cache entry gets recycled, we'd lose the failure reason. Alternatively, we could use a separate garbage collection thread to make sure every running pod keeps all their entries. I am not sure about adding the complexity since kubelet doesn't persist these kind of state (and it's best effort), but I can be convinced otherwise.

File: pkg/kubelet/kubelet_test.go [hunk 1]
```diff
@@ -564,7 +565,7 @@ func TestSyncPodsCreatesNetAndContainer(t *testing.T) {
 	waitGroup.Wait()
 
 	verifyCalls(t, fakeDocker, []string{
-		"list", "list", "list", "create", "start", "inspect_container", "create", "start", "list", "inspect_container", "inspect_container"})
+		"list", "list", "list", "inspect_image", "create", "start", "inspect_container", "create", "start", "list", "inspect_container", "inspect_container"})
```
Comment (dchen1107) [inline pkg/kubelet/kubelet_test.go]: Why extra inspect_image here?

**Comment by dchen1107** [comment]
LGTM except one question above.

File: pkg/kubelet/kubelet_test.go (hunk 1 above)
Comment (yujuhong) [inline pkg/kubelet/kubelet_test.go]: Before, if the pod infra container is not running and no user containers can be found, we'd return an error. Now, it'd continue to see if there is any creation error and inspect the images for the user containers.

Comment (dchen1107) [inline pkg/kubelet/kubelet_test.go]: ACK.

**Comment by dchen1107** [comment]
Thanks for fixing this. We have more comprehensive story on PodStatus now. Of course it is at the best effort. :-)

**Pull request merged**